File: dotnet/core/workspace/csharp/domain/core/derivations/default/Engine.cs
// <copyright file="DerivationFactory.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Derivations.Default
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Domain;
    using Meta;

    public class Engine
    {
        public IDictionary<Rule, ISet<IClass>> ClassesByRule { get; }

        public IDictionary<IClass, Rule[]> RulesByClass { get; }

        public IDictionary<IClass, IDictionary<IRoleType, ISet<RolePattern>>> PatternsByRoleTypeByClass { get; }

        public IDictionary<IClass, IDictionary<IAssociationType, ISet<AssociationPattern>>> PatternsByAssociationTypeByClass { get; }

        public IDictionary<Pattern, IRule> RuleByPattern { get; }

        public Engine(Rule[] rules)
        {
            this.ClassesByRule = new Dictionary<Rule, ISet<IClass>>();
            this.PatternsByRoleTypeByClass = new Dictionary<IClass, IDictionary<IRoleType, ISet<RolePattern>>>();
            this.PatternsByAssociationTypeByClass = new Dictionary<IClass, IDictionary<IAssociationType, ISet<AssociationPattern>>>();
            this.RuleByPattern = new Dictionary<Pattern, IRule>();

            foreach (var rule in rules)
            {
                var ruleClasses = new HashSet<IClass>();
                foreach (var pattern in rule.Patterns)
                {
                    this.RuleByPattern.Add(pattern, rule);

                    var patternClasses = pattern switch
                    {
                        RolePattern { ObjectType: null } rolePattern => rolePattern.RoleType.AssociationType.ObjectType.Classes.ToArray(),
                        RolePattern { ObjectType: { } } rolePattern => rolePattern.ObjectType.Classes.ToArray(),
                        AssociationPattern associationPattern => associationPattern.AssociationType.RoleType.ObjectType.IsComposite ? ((IComposite)associationPattern.AssociationType.RoleType.ObjectType).Classes.ToArray() : Array.Empty<IClass>(),
                        _ => Array.Empty<IClass>()
                    };

                    ruleClasses.UnionWith(patternClasses);

                    switch (pattern)
                    {
                        case RolePattern rolePattern:
                            foreach (var @class in patternClasses)
                            {
                                if (!this.PatternsByRoleTypeByClass.TryGetValue(@class, out var patternsByRoleType))
                                {
                                    patternsByRoleType = new Dictionary<IRoleType, ISet<RolePattern>>();
                                    this.PatternsByRoleTypeByClass.Add(@class, patternsByRoleType);
                                }

                                var roleType = rolePattern.RoleType;

                                if (!patternsByRoleType.TryGetValue(roleType, out var patterns))
                                {
                                    patterns = new HashSet<RolePattern>();
                                    patternsByRoleType.Add(roleType, patterns);
                                }

                                _ = patterns.Add(rolePattern);
                            }

                            break;
                        case AssociationPattern associationPattern:
                            foreach (var @class in patternClasses)
                            {
                                if (!this.PatternsByAssociationTypeByClass.TryGetValue(@class, out var patternsByAssociationType))
                                {
                                    patternsByAssociationType = new Dictionary<IAssociationType, ISet<AssociationPattern>>();
                                    this.PatternsByAssociationTypeByClass.Add(@class, patternsByAssociationType);
                                }

                                var associationType = associationPattern.AssociationType;

                                if (!patternsByAssociationType.TryGetValue(associationType, out var patterns))
                                {
                                    patterns = new HashSet<AssociationPattern>();
                                    patternsByAssociationType.Add(associationType, patterns);
                                }

                                _ = patterns.Add(associationPattern);
                            }

                            break;
                    }
                }

                this.ClassesByRule.Add(rule, ruleClasses);
            }

            var classes = this.ClassesByRule.SelectMany(v => v.Value).Distinct();
            this.RulesByClass = classes.ToDictionary(v => v, v => rules.Where(w => this.ClassesByRule[w].Contains(v)).ToArray());
        }
    }
}

File: typescript/modules/libs/workspace/adapters/system/src/lib/session/derivation/engine.ts
import { IAssociationPattern, IPattern, IRolePattern, IRule } from '@allors/workspace/domain/system';
import { AssociationType, Class, Composite, RoleType } from '@allors/workspace/meta/system';

export class Engine {
  classesByRule: Map<IRule, Set<Class>>;

  rulesByClass: Map<Class, IRule[]>;

  patternsByRoleTypeByClass: Map<Class, Map<RoleType, Set<IRolePattern>>>;

  patternsByAssociationTypeByClass: Map<Class, Map<AssociationType, Set<IAssociationPattern>>>;

  ruleByPattern: Map<IPattern, IRule>;

  patternsByClassByRule: Map<IRule, Map<Class, IPattern[]>>;

  public constructor(public rules: Readonly<IRule[]>) {
    Object.freeze(this.rules);
    this.classesByRule = new Map();
    this.rulesByClass = new Map();
    this.patternsByRoleTypeByClass = new Map();
    this.patternsByAssociationTypeByClass = new Map();
    this.ruleByPattern = new Map();
    this.patternsByClassByRule = new Map();

    for (const rule of rules) {
      let ruleClasses = new Set<Class>();

      for (const pattern of rule.patterns) {
        this.ruleByPattern.set(pattern, rule);

        let patternClasses: Set<Class>;

        switch (pattern.kind) {
          case 'RolePattern':
            if (pattern.objectType) {
              patternClasses = pattern.objectType.classes;
            } else {
              patternClasses = (pattern.roleType.associationType.objectType as Composite).classes;
            }
            break;
          case 'AssociationPattern':
            if (pattern.AssociationType.roleType.objectType.isComposite) {
              ruleClasses = new Set([...ruleClasses, ...(pattern.AssociationType.roleType.objectType as Composite).classes]);
            }
            break;
        }

        if (patternClasses) {
          ruleClasses = new Set([...ruleClasses, ...patternClasses]);

          let patternsByClass = this.patternsByClassByRule.get(rule);
          if (patternsByClass == null) {
            patternsByClass = new Map();
            this.patternsByClassByRule.set(rule, patternsByClass);
          }

          for (const patternClass of patternClasses) {
            let patterns = patternsByClass.get(patternClass);
            if (patterns == null) {
              patterns = [];
              patternsByClass.set(patternClass, patterns);
            }

            patterns.push(pattern);
          }

          switch (pattern.kind) {
            case 'RolePattern':
              for (const patternClass of patternClasses) {
                let patternsByRoleType = this.patternsByRoleTypeByClass.get(patternClass);
                if (patternsByRoleType == null) {
                  patternsByRoleType = new Map();
                  this.patternsByRoleTypeByClass.set(patternClass, patternsByRoleType);
                }

                const roleType = pattern.roleType;

                let patterns = patternsByRoleType.get(roleType);
                if (patterns == null) {
                  patterns = new Set();
                  patternsByRoleType.set(roleType, patterns);
                }

                patterns.add(pattern);
              }

              break;

            case 'AssociationPattern':
              for (const patternClass of patternClasses) {
                let patternsByAssociationType = this.patternsByAssociationTypeByClass.get(patternClass);
                if (patternsByAssociationType == null) {
                  patternsByAssociationType = new Map();
                  this.patternsByAssociationTypeByClass.set(patternClass, patternsByAssociationType);
                }

                const AssociationType = pattern.AssociationType;

                let patterns = patternsByAssociationType.get(AssociationType);
                if (patterns == null) {
                  patterns = new Set();
                  patternsByAssociationType.set(AssociationType, patterns);
                }

                patterns.add(pattern);
              }

              break;
          }
        }
      }

      this.classesByRule.set(rule, ruleClasses);
    }

    const classes = new Set<Class>();
    for (const [, ruleClasses] of this.classesByRule) {
      for (const ruleClass of ruleClasses) {
        classes.add(ruleClass);
      }
    }

    for (const cls of classes) {
      const classRules = rules.filter((v) => this.classesByRule.get(v)?.has(cls));
      this.rulesByClass.set(cls, classRules);
    }
  }
}
